numpy/_core/src/multiarray/unique.cpp

```diff
@@ -0,0 +1,180 @@
+#define NPY_NO_DEPRECATED_API NPY_API_VERSION
+#define _MULTIARRAYMODULE
+
+#include <Python.h>
+
+#include <unordered_set>
+#include <functional>
+
+#include <numpy/npy_common.h>
+#include "numpy/arrayobject.h"
+
+// This is to use RAII pattern to handle cpp exceptions while avoiding memory leaks.
+// Adapted from https://stackoverflow.com/a/25510879/2536294
+template <typename F>
+struct FinalAction {
+    FinalAction(F f) : clean_{f} {}
+    ~FinalAction() { clean_(); }
+  private:
+    F clean_;
+};
+
+template <typename F>
+FinalAction<F> finally(F f) {
+    return FinalAction<F>(f);
+}
+
+template<typename T>
+static PyObject*
+unique(PyArrayObject *self)
+{
+    /* This function takes a numpy array and returns a numpy array containing
+    the unique values.
+
+    It assumes the numpy array includes data that can be viewed as unsigned integers
+    of a certain size (sizeof(T)).
+
+    It doesn't need to know the actual type, since it needs to find unique values
+    among binary representations of the input data. This means it won't apply to
+    custom or complicated dtypes or string values.
+    */
+    NPY_ALLOW_C_API_DEF;
+    std::unordered_set<T> hashset;
+
+    NpyIter *iter = NpyIter_New(self, NPY_ITER_READONLY |
+                                      NPY_ITER_EXTERNAL_LOOP |
+                                      NPY_ITER_REFS_OK |
+                                      NPY_ITER_ZEROSIZE_OK |
+                                      NPY_ITER_GROWINNER,
+                                NPY_KEEPORDER, NPY_NO_CASTING,
+                                NULL);
+    // Making sure the iterator is deallocated when the function returns, with
+    // or w/o an exception
+    auto iter_dealloc = finally([&]() { NpyIter_Deallocate(iter); });
+    if (iter == NULL) {
+        return NULL;
+    }
+
+    NpyIter_IterNextFunc *iternext = NpyIter_GetIterNext(iter, NULL);
+    if (iternext == NULL) {
+        return NULL;
+    }
+    char **dataptr = NpyIter_GetDataPtrArray(iter);
+    npy_intp *strideptr = NpyIter_GetInnerStrideArray(iter);
+    npy_intp *innersizeptr = NpyIter_GetInnerLoopSizePtr(iter);
+
+    // release the GIL
+    PyThreadState *_save;
+    _save = PyEval_SaveThread();
+    // Making sure the GIL is re-acquired when the function returns, with
+    // or w/o an exception
+    auto grab_gil = finally([&]() { PyEval_RestoreThread(_save); });
+    // first we put the data in a hash map
+    do {
+        char* data = *dataptr;
+        npy_intp stride = *strideptr;
+        npy_intp count = *innersizeptr;
+
+        while (count--) {
+            hashset.insert(*((T *) data));
+            data += stride;
+        }
+    } while(iternext(iter));
+
+    npy_intp length = hashset.size();
+
+    NPY_ALLOW_C_API;
+    PyArray_Descr *descr = PyArray_DESCR(self);
+    Py_INCREF(descr);
+    PyObject *res_obj = PyArray_NewFromDescr(
```

Comment:
If there was a C++ exception after this, we would leak it... But, I don't think that is possible, so let's not worry about it.